src/stream/buffer_unordered.rs

```diff
@@ -54,25 +62,31 @@ impl<S> Stream for BufferUnordered<S>
         // Next, try and step futures forward until we find a ready one.
         // Always start at `cur` for fairness.
         let mut waiting = false;
-        for i in 0..self.futures.len() {
-            let mut idx = self.cur + i;
-            if idx >= self.futures.len() {
-                idx -= self.futures.len();
-            }
+        let mut drained = self.stack.drain();
+        while let Some(idx) = drained.next() {
             let future = &mut self.futures[idx];
             let result = match *future {
-                Some(ref mut s) => match s.poll() {
-                    Ok(Async::NotReady) => {
-                        waiting = true;
-                        continue
-                    },
-                    Ok(Async::Ready(e)) => Ok(Async::Ready(Some(e))),
-                    Err(e) => Err(e),
+                Some(ref mut s) => {
+                    let event = UnparkEvent::new(self.stack.clone(), idx);
+                    let poll_result = task::with_unpark_event(event, || {
+                        s.poll()
+                    });
+                    match poll_result {
+                        Ok(Async::NotReady) => {
+                            waiting = true;
+                            continue
+                        },
+                        Ok(Async::Ready(e)) => Ok(Async::Ready(Some(e))),
+                        Err(e) => Err(e),
+                    }
                 },
                 None => continue,
             };
-            self.cur = i + 1;
             *future = None;
+            // push unresolved futures back
+            for idx in drained {
+                self.stack.push(idx);
```

Comment:
This may also be related to the scalability problem, this means that if there are N ready connections that we'll do N^2 operations to actually process them I think. It should be possible to implement an O(1) append I think with the stack here, though?